homeassistant/components/nissan_leaf/__init__.py

```diff
@@ -0,0 +1,514 @@
+CONF_FORCE_MILES = 'force_miles'
+
+INITIAL_UPDATE = timedelta(seconds=15)
+MIN_UPDATE_INTERVAL = timedelta(minutes=2)
+DEFAULT_INTERVAL = timedelta(hours=1)
+DEFAULT_CHARGING_INTERVAL = timedelta(minutes=15)
+DEFAULT_CLIMATE_INTERVAL = timedelta(minutes=5)
+RESTRICTED_BATTERY = 2
+RESTRICTED_INTERVAL = timedelta(hours=12)
+
+MAX_RESPONSE_ATTEMPTS = 10
+
+PYCARWINGS2_SLEEP = 30
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.All(cv.ensure_list, [vol.Schema({
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Required(CONF_REGION): vol.In(CONF_VALID_REGIONS),
+        vol.Optional(CONF_NCONNECT, default=True): cv.boolean,
+        vol.Optional(CONF_INTERVAL, default=DEFAULT_INTERVAL): (
+            vol.All(cv.time_period, vol.Clamp(min=MIN_UPDATE_INTERVAL))),
+        vol.Optional(CONF_CHARGING_INTERVAL,
+                     default=DEFAULT_CHARGING_INTERVAL): (
+                         vol.All(cv.time_period,
+                                 vol.Clamp(min=MIN_UPDATE_INTERVAL))),
+        vol.Optional(CONF_CLIMATE_INTERVAL,
+                     default=DEFAULT_CLIMATE_INTERVAL): (
+                         vol.All(cv.time_period,
+                                 vol.Clamp(min=MIN_UPDATE_INTERVAL))),
+        vol.Optional(CONF_FORCE_MILES, default=False): cv.boolean
+    })])
+}, extra=vol.ALLOW_EXTRA)
+
+LEAF_COMPONENTS = [
+    'sensor', 'switch', 'binary_sensor', 'device_tracker'
+]
+
+SIGNAL_UPDATE_LEAF = 'nissan_leaf_update'
+
+SERVICE_UPDATE_LEAF = 'update'
+ATTR_VIN = 'vin'
+
+UPDATE_LEAF_SCHEMA = vol.Schema({
+    vol.Required(ATTR_VIN): cv.string,
+})
+
+
+async def async_setup(hass, config):
+    """Set up the Nissan Leaf component."""
+    import pycarwings2
+
+    async def handle_update(service):
+        # It would be better if this was changed to use nickname, or
+        # an entity name rather than a vin.
+        vin = service.data.get(ATTR_VIN, '')
+
+        if vin in hass.data[DATA_LEAF]:
+            data_store = hass.data[DATA_LEAF][vin]
+            async_track_point_in_utc_time(
+                hass, data_store.async_update_data, utcnow())
+            return True
+
+        _LOGGER.debug("Vin %s not recognised for update", vin)
+        return False
+
+    async def async_setup_leaf(car_config):
+        """Set up a car."""
+        _LOGGER.debug("Logging into You+Nissan...")
+
+        username = car_config[CONF_USERNAME]
+        password = car_config[CONF_PASSWORD]
+        region = car_config[CONF_REGION]
+        leaf = None
+
+        async def leaf_login():
+            nonlocal leaf
+            sess = pycarwings2.Session(username, password, region)
+            leaf = sess.get_leaf()
+
+        try:
+            # This might need to be made async (somehow) causes
+            # homeassistant to be slow to start
+            await hass.async_add_job(leaf_login)
+        except(RuntimeError, urllib.error.HTTPError):
```

Comment:
`RuntimeError` is a broad exception we should avoid to catch. Why do we need to do that?